File: compiler/nodes/patterns/unit.go
package patterns

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type UnitPatternNode struct {
	Facts *database.Facts

	matching database.Node
}

func (node *UnitPatternNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseUnitPattern(parser *syntax.Parser) (*UnitPatternNode, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("LeftParenthesis", syntax.TokenConfig{
		Reason: "between these parentheses",
	})
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("RightParenthesis")
	if err != nil {
		return nil, err
	}

	return &UnitPatternNode{
		Facts: database.NewFacts(span()),
	}, nil
}

func (node *UnitPatternNode) Visit(visitor *visit.Visitor) {
	visitPattern(visitor, node)
	node.matching = visitor.CurrentMatch.Node

	visitor.Constraint(typecheck.TypeConstraint(node, typecheck.TupleType([]typecheck.Type{})))
}

func (node *UnitPatternNode) Codegen(c *codegen.Codegen) error {
	// No code needed
	return nil
}

File: compiler/src/nodes/patterns/unit.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenError},
    database::{Node, NodeRef},
    nodes::visit_pattern,
    syntax::{ParseError, Parser, TokenKind},
    typecheck::TypeConstraint,
    visit::{Visit, Visitor},
};

#[derive(Debug)]
pub struct UnitPatternNode;

impl Node for UnitPatternNode {}

pub fn parse_unit_pattern(parser: &mut Parser<'_>) -> Result<UnitPatternNode, ParseError> {
    parser.token_with_reason(TokenKind::LeftParenthesis, "between these parentheses")?;
    parser.token(TokenKind::RightParenthesis)?;

    Ok(UnitPatternNode)
}

impl Visit for UnitPatternNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_pattern(node, visitor);

        visitor.constraint(TypeConstraint::new(node.clone(), visitor.unit_type()));
    }
}

impl Codegen for UnitPatternNode {
    fn codegen(&self, _codegen: &mut CodegenCtx<'_>) -> Result<(), CodegenError> {
        // No code needed
        Ok(())
    }
}
